File: apps/web/src/features/career/_types/match.ts
import type { Lineup, Tactic } from "./tactics";

export type PlayRequest = { seed: number; homeLineup: Lineup; homeTactic: Tactic };
export type MatchEvent = { sequence: number; minute: number; second: number; type: string; team_id: string | null; player_id: string | null; zone: string | null; payload: Record<string, string | number | boolean | null> };
export type TeamStats = { team_id: string; goals: number; shots: number; shots_on_target: number; xg: number; possession: number; passes_attempted: number; passes_completed: number; fouls: number; yellow_cards: number; red_cards: number };
export type PlayerStats = { player_id: string; team_id: string; minutes: number; rating: number; goals: number; assists: number; shots: number; passes_attempted: number; passes_completed: number; tackles: number };
export type MatchResult = { seed: number; engine_version: string; ruleset_version: string; home_team_id: string; away_team_id: string; home_score: number; away_score: number; events: MatchEvent[]; stats: { home: TeamStats; away: TeamStats; players: PlayerStats[] } };
export type PlayResponse = { matchId: string; result: MatchResult; matchdayNumber: number; simulatedAiMatchIds: string[]; failedFixtureIds: string[]; matchdayComplete: boolean };
export type MatchPauseReason = "KICKOFF" | "KEY_HIGHLIGHT" | "MILESTONE" | "HALF_TIME" | "INJURY" | "TACTICAL_DECISION" | "FULL_TIME";
export type MatchSessionTeam = { clubId?: string; id?: string; clubName?: string; name?: string; lineup?: Lineup; tactic?: Tactic; inactivePlayerIds?: string[] };
export type MatchSessionSnapshot = {
  id: string; fixtureId: string; matchId?: string | null; status: "ACTIVE" | "COMPLETED" | "ABANDONED" | string;
  version: number; minute: number; score?: { home: number; away: number }; homeScore?: number; awayScore?: number;
  pauseReason: MatchPauseReason | string; controlledClubId?: string; home?: MatchSessionTeam; away?: MatchSessionTeam;
  events?: MatchEvent[]; eventsSinceLastPause?: MatchEvent[];
  stats?: { home: Omit<TeamStats, "team_id"> & { team_id?: string }; away: Omit<TeamStats, "team_id"> & { team_id?: string }; players?: PlayerStats[] };
  substitutions?: { playersUsed: number; playersRemaining: number; windowsUsed: number; windowsRemaining: number; halftimeWindowFree: boolean };
  canContinue?: boolean; canFinish?: boolean;
};
export type MatchSessionAction = { requestId: string; expectedVersion: number };
export type MatchSessionCommandPayload =
  | { type: "TACTIC"; tactic: Tactic; lineup?: Lineup }
  | { type: "SHOUT"; shout: "ENCOURAGE" | "DEMAND_MORE" | "FOCUS" | "CALM_DOWN" }
  | { type: "SUBSTITUTION"; substitutions: { outgoingPlayerId: string; incomingPlayerId: string }[] };
export type MatchSessionCommand = MatchSessionAction & MatchSessionCommandPayload;
export type StartMatchSessionRequest = { requestId: string; seed?: number; lineup?: Lineup; tactic?: Tactic };

File: services/match-engine/match_engine/domain/match.py
from typing import Annotated
from uuid import UUID
from pydantic import Field, JsonValue, model_validator

from .base import DomainModel, Percentage, Version
from .enums import EventType, Zone
from .team import TeamSnapshot


class MatchInput(DomainModel):
    seed: Annotated[int, Field(ge=0, le=2**63 - 1)]
    engine_version: Version
    ruleset_version: Version
    home: TeamSnapshot
    away: TeamSnapshot

    @model_validator(mode="after")
    def different_teams(self):
        if self.home.id == self.away.id:
            raise ValueError("Home and away teams must be different")
        home_players = {player.id for player in self.home.players}
        away_players = {player.id for player in self.away.players}
        if home_players & away_players:
            raise ValueError("Home and away rosters cannot share players")
        return self


class MatchEvent(DomainModel):
    sequence: Annotated[int, Field(ge=1)]
    minute: Annotated[int, Field(ge=0, le=130)]
    second: Annotated[int, Field(ge=0, le=59)] = 0
    type: EventType
    team_id: UUID | None = None
    player_id: UUID | None = None
    zone: Zone | None = None
    payload: dict[str, JsonValue] = Field(default_factory=dict)


class TeamStats(DomainModel):
    team_id: UUID
    goals: Annotated[int, Field(ge=0)] = 0
    shots: Annotated[int, Field(ge=0)] = 0
    shots_on_target: Annotated[int, Field(ge=0)] = 0
    xg: Annotated[float, Field(ge=0)] = 0
    possession: Percentage = 50
    passes_attempted: Annotated[int, Field(ge=0)] = 0
    passes_completed: Annotated[int, Field(ge=0)] = 0
    fouls: Annotated[int, Field(ge=0)] = 0
    yellow_cards: Annotated[int, Field(ge=0)] = 0
    red_cards: Annotated[int, Field(ge=0)] = 0

    @model_validator(mode="after")
    def valid_totals(self):
        if self.shots_on_target > self.shots:
            raise ValueError("Shots on target cannot exceed shots")
        if self.passes_completed > self.passes_attempted:
            raise ValueError("Completed passes cannot exceed attempted passes")
        return self


class PlayerStats(DomainModel):
    player_id: UUID
    team_id: UUID
    minutes: Annotated[int, Field(ge=0, le=130)] = 0
    rating: Annotated[float, Field(ge=1, le=10)] = 6
    goals: Annotated[int, Field(ge=0)] = 0
    assists: Annotated[int, Field(ge=0)] = 0
    shots: Annotated[int, Field(ge=0)] = 0
    passes_attempted: Annotated[int, Field(ge=0)] = 0
    passes_completed: Annotated[int, Field(ge=0)] = 0
    tackles: Annotated[int, Field(ge=0)] = 0


class MatchStats(DomainModel):
    home: TeamStats
    away: TeamStats
    players: tuple[PlayerStats, ...] = ()

    @model_validator(mode="after")
    def valid_stats(self):
        if self.home.team_id == self.away.team_id:
            raise ValueError("Stats must belong to two different teams")
        if abs(self.home.possession + self.away.possession - 100) > 0.1:
            raise ValueError("Team possession must total 100")
        player_ids = [player.player_id for player in self.players]
        if len(set(player_ids)) != len(player_ids):
            raise ValueError("Player stats must be unique")
        return self


class MatchResult(DomainModel):
    seed: Annotated[int, Field(ge=0, le=2**63 - 1)]
    engine_version: Version
    ruleset_version: Version
    home_team_id: UUID
    away_team_id: UUID
    home_score: Annotated[int, Field(ge=0)]
    away_score: Annotated[int, Field(ge=0)]
    events: tuple[MatchEvent, ...]
    stats: MatchStats

    @model_validator(mode="after")
    def internally_consistent(self):
        if self.home_team_id == self.away_team_id:
            raise ValueError("Home and away teams must be different")
        if self.stats.home.team_id != self.home_team_id or self.stats.away.team_id != self.away_team_id:
            raise ValueError("Stats team IDs do not match result team IDs")
        if self.stats.home.goals != self.home_score or self.stats.away.goals != self.away_score:
            raise ValueError("Stats goals do not match result score")
        if [event.sequence for event in self.events] != list(range(1, len(self.events) + 1)):
            raise ValueError("Event sequence must be contiguous and start at 1")
        return self


class TeamSimulationState(DomainModel):
    possessions: int = 0
    goals: int = 0
    shots: int = 0
    shots_on_target: int = 0
    xg: float = 0
    passes_attempted: int = 0
    passes_completed: int = 0
    fouls: int = 0
    yellow_cards: int = 0
    red_cards: int = 0


class MatchSimulationState(DomainModel):
    minute: Annotated[int, Field(ge=1, le=131)] = 1
    rng_state: str
    home: TeamSimulationState = Field(default_factory=TeamSimulationState)
    away: TeamSimulationState = Field(default_factory=TeamSimulationState)
    current_home: TeamSnapshot
    current_away: TeamSnapshot
    player_data: dict[str, dict[str, int]] = Field(default_factory=dict)
    minutes: dict[str, int] = Field(default_factory=dict)
    events: tuple[MatchEvent, ...] = ()
    reacted: tuple[str, ...] = ()
    half_time_added: bool = False
    substitutions_added: bool = False
    completed: bool = False
